enhancements/installer/feature-sets.md

```diff
@@ -0,0 +1,209 @@
+		errMsg := "the TechPreviewNoUpgrade feature set must be enabled to use this field"
+
+    if installConfig.Foo != nil && len(installConfig.Foo.Bar) > 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("foo", "bar"), errMsg))
+		}
+	}
+```
+
+Package `types` defines the install config and represents the API offered by the Installer. Most notably this
+API is consumed/used by hive. For this reason, it is important that feature sets be part of the install config.
+
+#### CLI Flag (future work)
+
+On the other hand, the install config is an asset and gated features could potentially be exercised
+before the install config is validated (or perhaps the feature is completely independent from the install config). For this reason, it would
+be a good idea to provide a flag:
+
+```shell
+$ openshift-install create cluster --feature-set TechPreviewNoUpgrade
+```
+
+This flag, if provided, would also populate the value in the install config; and conflicts between the flag and install config
+would throw an error.
+
+While there are clear use cases for a flag, there are no immediate use cases and the implementation of the Installer (with
+its separation between command-line flags and assets) does not lead to a clear-cut design. Implementing this functionality
+is certainly possible, but I propose we defer implementation for this until it is determined to be necessary.
+
+#### openshift-install explain
+
+`explain` provides a command-line reference for the install config:
+
+```shell
+$ openshift-install explain installconfig | tail -n 6
+    pullSecret <string> -required-
+      PullSecret is the secret to use when pulling images.
+
+    sshKey <string>
+      SSHKey is the public Secure Shell (SSH) key to provide access to instances.
+```
+
+`explain` should handle feature-gated fields appropriately, either by skipping them (not printing them) or clearly marking
+them as gated features.
+
+### Risks and Mitigations
+
+I am not aware of any significant risks to introducing this feature. 
+
+### Drawbacks
+
+I am not aware of any significant drawbacks to introducing this feature.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. The [Tech Preview Guidelines](https://github.com/openshift/enhancements/blob/master/guidelines/techpreview.md)
+state that Tech Preview features, and perhaps all gated features, need to be listed in the [OpenShift API](https://github.com/openshift/api/blob/bace76a807222b30bb9bfd4926826348156fb522/config/v1/types_feature.go#L117). Is that necessary for the Installer, or
+only for operators?
+
+### Test Plan
+
+Unit testing the new tech preview functionality and observing existing e2e tests should be sufficient.
+
+### Graduation Criteria
+
+n/a
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+
+### Upgrade / Downgrade Strategy
+
+Upgrades will be blocked.
+
+### Version Skew Strategy
+
+n/a
```

Comment:
Released installers are linked to a specific installed version, but that's not always quite true in development/CI.  There seems like there's a small risk in skew between a feature-set change in openshift/api, that change making its way into the cluster (via openshift/cluster-config-operator bumping its openshift/api vendor), and that change making its way into the installer (via openshift/installer bumping its openshift/api vendor).  I dunno if it's worth digging into these skew issues in the enhancement, but might be worth acknowledging them, and then saying that we expect the skew to be infrequent (because feature sets are rarely added or removed) and also short (because anyone who touches them will remember to vendor-bump both the installer and config-operator after landing the openshift/api bump).  Or maybe  it is worth trying to find some way to remind folks to vendor-bump openshift/api in the installer?  Or something?

Reply:
Added to the most recent update.